Add data structure for memory efficient lists of std::variants

weinig opened the pull request:
#### fdc91dd3319dac7f95a23885797a39c7951a5a6c
<pre>
Add data structure for memory efficient lists of std::variants
<a href="https://bugs.webkit.org/show_bug.cgi?id=283423">https://bugs.webkit.org/show_bug.cgi?id=283423</a>

Reviewed by Darin Adler.

Adds a new data structure, VariantList&lt;std::variant&lt;Ts...&gt;&gt; which acts
like a Vector&lt;std::variant&lt;Ts...&gt;&gt; except that each element only takes up
its own size (+ tag size). The tradeoff is that the list is not random
access capable, and access to the elements is only available via iteration.
In practice, this tradeoff is a good one for the CSS and Style value types.

Rather than vending std::variants, VariantList vends VariantListProxy
objects which act like std::variants, but don&apos;t require the extra costs
of construction.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/VariantList.h: Added.
* Source/WTF/wtf/VariantListOperations.h: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/VariantList.cpp: Added.

Canonical link: <a href="https://commits.webkit.org/287010@main">https://commits.webkit.org/287010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ece8d1b1245f856302cebd90f64b1ac9336a812

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77908 "24 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82556 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80029 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5236 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61006 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18937 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80976 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50993 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66925 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41309 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48360 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24377 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27508 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71052 "Found 2 new JSC stress test failures: stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-no-inline-validate, stress/get-by-val-hoist-above-structure-2.js.mini-mode (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69454 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24730 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83918 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77143 "Found 2 new JSC stress test failures: stress/get-by-val-hoist-above-structure.js.eager-jettison-no-cjit, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-no-cjit (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5275 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3581 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69225 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-with-will-change-transform-001.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-skew.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5431 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68481 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17109 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10601 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99456 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5223 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21715 "Found 2 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5215 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7000 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -155,6 +155,36 @@ inline bool is8ByteAligned(void* p)
     return !((uintptr_t)(p) & (sizeof(double) - 1));
 }
 
+inline std::byte* alignedBytes(std::byte* pointer, size_t alignment)
```
For the most part, @cdumez has convinced me to just use `uint8_t` for generic bytes and not start to use `std::byte`. Maybe you can convince me otherwise, after all `uint8_t` and `LChar` are the same type and one advantage of `std::byte` is that it’s not the same as any other type!

weinig replied:
Other than some extra overloads, what's the argument?

darinadler replied:
I guess maybe extra type conversions?

weinig replied:
I would think that would be just a temporary problem while making the move from uint8_t to std::byte for actual opaque byte buffers. 

But either way, this is entirely constrained to the data structure, so if we decide we really don't like std::byte, changing it will be very easy.

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -832,6 +862,18 @@ T& reinterpretCastSpanStartTo(std::span<uint8_t, Extent> span)
 }
 
+template<typename T, std::size_t Extent>
+const T& reinterpretCastSpanStartTo(std::span<const std::byte, Extent> span)
```
This additional overloading is part of the cost of starting to use `std::byte`. I suppose `byteCast` sprinkling is less elegant.

darinadler commented on Source/WTF/wtf/VariantList.h:
```diff
@@ -0,0 +1,354 @@
+namespace WTF {
+
+template<typename Variant, size_t inlineCapacity = 0> class VariantList;
+template<typename Variant, size_t inlineCapacity = 0> struct VariantListConstIterator;
```
Consider putting these into `Forward.h` instead?

darinadler commented on Source/WTF/wtf/VariantListOperations.h:
```diff
@@ -0,0 +1,527 @@
+using VariantListTag = unsigned;
+
+template<typename Variant> struct VariantListProxy;
+template<typename Variant> struct VariantListSizer;
```
I would normally omit the type argument name `Variant` from these forward declarations, but I can see including it if you think it adds clarity. I would also put forward declarations like these above a type definition like `VariantListTag`.

darinadler commented on Source/WTF/wtf/VariantListOperations.h:
```diff
@@ -0,0 +1,527 @@
+        }                                                                                           \
+
+    switch (tag) {
+        WTF_VARIANT_LIST_VISIT_CASE(0, WTF_VARIANT_LIST_VISIT_CASE_COUNT)
```
An idiom like this seems like it needs the comment that explains what its intentions are.

weinig committed:
Add data structure for memory efficient lists of std::variants
https://bugs.webkit.org/show_bug.cgi?id=283423

Reviewed by Darin Adler.

Adds a new data structure, VariantList<std::variant<Ts...>> which acts
like a Vector<std::variant<Ts...>> except that each element only takes up
its own size (+ tag size). The tradeoff is that the list is not random
access capable, and access to the elements is only available via iteration.
In practice, this tradeoff is a good one for the CSS and Style value types.

Rather than vending std::variants, VariantList vends VariantListProxy
objects which act like std::variants, but don't require the extra costs
of construction.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/VariantList.h: Added.
* Source/WTF/wtf/VariantListOperations.h: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/VariantList.cpp: Added.

Canonical link: https://commits.webkit.org/287010@main
